engine/src/engine.ts

```diff
@@ -0,0 +1,181 @@
+import Board from './board'
+import Player from './player'
+
+interface EngineConfig {
+  boardSize: {
+    rows: number
+    columns: number
+  }
+  winLength: number
+  players: Player[]
+  firstPlayer: Player
+}
+
+export default class Engine {
+  private board: Board
+  private turnPlayer: Player
+  private config: EngineConfig
+
+  constructor(config: EngineConfig) {
+    if (config.players.length !== 2) {
+      throw new Error('Invalid number of players')
+    } else if (config.players[0].value === config.players[1].value) {
+      throw new Error('Players should have different values')
+    } else if (
+      !config.players.find((player) => player === config.firstPlayer)
+    ) {
+      throw new Error('First player should be one of the players')
+    }
+
+    this.board = new Board()
+    this.turnPlayer = config.firstPlayer
+    this.config = config
+  }
+
+  private init() {
+    const { boardSize } = this.config
+    this.board.init(boardSize.rows, boardSize.columns)
+  }
+
+  public start() {
+    this.init()
+  }
+
+  public getTurnPlayer() {
+    return this.turnPlayer
+  }
+
+  public play(player: Player, row: number, column: number) {
+    // check if the player is valid
+    if (!this.config.players.find((p) => p === player)) {
```

Comment:
For this and the following sanity checks, use console.assert(), making it clear that a violation would be an internal error.

Reply:
I think this is more of a check for external error, because the client (CLI or Web) may try to make play for the wrong user